Initial accessibility support in Compose

Leedwon opened the pull request:
## Description

This PR adds minimal accessibility support to Vico charts. It introduces a `contentDescription` parameter, allowing meaningful descriptions to be passed for data points. These descriptions will be read by TalkBack, improving the chart experience for screen reader users.

Note: this doesn’t make the entire chart accessible, full support should be introduced in separate PRs. If preferred, I can retarget this PR to a feature branch instead of master, but it should already provide some accessibility improvements on its own.

How to Review This PR:
The suggested way to review this PR is by going commit-by-commit. I’ve left comments in places where I saw potential for improvement or ran into limitations. I hope that your deep knowledge of the library can help improve or fix some of these areas.

## PR Scope
- Chart data points can now include a content description that will be announced by TalkBack, making them more accessible to screen reader users.

## Out of scope
- Chart legend accessibility
- Chart markers accessibility
- Chart decorations accessibility
- Automatically scrolling to chart elements that are out of screen bounds as accessibility focus moves.
This should be handled in a future PR, as it requires implementing `BringIntoViewRequester` together with `BringIntoViewResponder` and connecting it to `VicoScrollState`. The main challenge here is the lack of a reliable callback that tells us when an element is focused by accessibility services unfortunately, `.onFocusChanged { }` doesn’t get triggered during accessibility navigation. To keep this PR smaller, I’m leaving this out for now.

## Demos

In each demo, TalkBack is enabled and I’m using left/right swipe gestures (default accessibility navigation method), to move between elements with semantics. You’ll notice that chart data points with `contentDescription` are now announced.

Previously, the entire chart was skipped, and accessibility focus jumped directly to previous/next arrows.

Please watch the demos with sound on to hear the TalkBack announcements in action.

| Chart Type         | Accessibility Demo |
|--------------------|--------------------|
| Column Chart       | <video src="https://github.com/user-attachments/assets/d84a3999-bced-4b35-a5d9-2bf6e959790d" width="450" controls></video> |
| Line Chart         | <video src="https://github.com/user-attachments/assets/d7634244-3147-4a51-b493-c64d3c40bee6" width="450" controls></video> |
| Candlestick Chart  | <video src="https://github.com/user-attachments/assets/63b1f2ec-8150-44a6-8f10-95f880828b5c" width="450" controls></video> |

## Known issues

This section describes known issues currently affecting chart accessibility. These can be addressed in this PR if you consider them blockers, or in future PRs. In most cases, I ran into some issues while trying to fix them, which is why they’re currently left as-is.

### Combo charts

In combo charts, some data points may prevent others from being announced. For example, in the demo video, you can see that for x = 8 and x = 9, the column data points are not announced. This happens because the line points are positioned higher than the columns, and the accessibility system assumes the columns are not visible (and thus not focusable).

https://github.com/user-attachments/assets/fb516788-ca4f-4922-9e2c-70b005f2d546

### Negative values

The highlighter for negative values starts from the bottom instead of the top. It might look a bit weird visually, but it’s not a problem for TalkBack.

https://github.com/user-attachments/assets/63d3a5ab-4255-4fe3-81a9-5396553077a8

### Clickable charts with markers

When charts contain markers (making them clickable), the accessibility focus might jump to the touched position. This only happens when the user swipes back and forth through accessibility elements while swiping on the chart itself. If the user swipes just under or over the chart, focus moves correctly between data points.. This can be seen i.e on the Electric car sales chart.

## Resources
https://eevis.codes/blog/2023-07-24/more-accessible-graphs-with-jetpack-compose-part-1-adding-content/

Leedwon committed:
Initial accessibility support in Compose

Co-authored-by: Patryk Goworowski <patrykgoworowski@gmail.com>
Co-authored-by: Patrick Michalik <120058021+patrickmichalik@users.noreply.github.com>

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,159 @@
+  Box(
+    modifier =
+      Modifier.offset(x = canvasX.pxToDp() - width / 2, y = canvasY.pxToDp())
+        .border(width = borderWidth, color = Color.Transparent)
```
This creates a transparent box that isn’t visible or interactable, but since it has semantics, it can be accessed via accessibility services.

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,159 @@
+) {
+  val width = xSpacing.pxToDp()
+
+  Box(
```
Since boxes can be layered on top of each other, the issue with combo charts applies (See PR description for details). One possible workaround could be to create very small boxes (e.g. 1.dp in size). This would likely fix the issue and ensure everything gets announced properly in combo charts.

However, it would remove the visual highlight of the current section, and overall it felt a bit hacky to me. Let me know what you think.

Gowsky replied:
Resolved by grouping by _x_ value, as discussed.

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -174,59 +175,66 @@ internal fun CartesianChartHostImpl(
-                  zoomState.zoom(factor, centroid.x, scrollState.value, layerBounds.value).let {
-                    scroll ->
-                    coroutineScope.launch { scrollState.scroll(scroll) }
+  Box {
```
A little bit weird git diff in here - the only actual change was wrapping the Canvas in a Box and displaying AccessibilityHighlighter alongside it. Canvas content itself hasn't changed

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -207,6 +210,7 @@ internal fun CartesianChartHostImpl(
 
       layerDimensions.clear()
       chart.prepare(measuringContext, layerDimensions)
+      targets = chart.allTargets
```
We need to update `targets` after calling `chart.prepare()`, since it looks like they're updated during that call - before it, `targets` were empty.

Let me know if I got that right, or if there's a better place to do this and it works differently under the hood.

Gowsky replied:
`targets` are updated in `CartesianChart.draw` call. Also, it seems that we don't need `targets` `MutableState` property. The targets can be passed directly to `AccessibilityHighlighter` from `CartesianChart`.

Leedwon replied:
Without `targets` `MutableState` the `AccessibilityHighlighter` is never recomposed with the updated `targets`. I initially used the approach that you mention - passing targets directly from `CartesianChart`, but it only passes initial targets which are empty list. This can be tested if you replace `targets = targets` with `targets = chart.allTargets`. I'll focus on other improvements first and we can improve this once we have the Provider approach in place

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -235,11 +238,13 @@ internal fun CartesianChartHostImpl(
-      canvasHeight = canvasBounds.height(),
-      xSpacing = layerDimensions.xSpacing,
-    )
+    if (isTouchExplorationEnabled()) {
```
I'm not 100% sure about this change. It’s kind of a safeguard to avoid doing extra work when accessibility isn’t enabled — but at the same time, it feels like we’re treating accessibility as a second-class citizen. Curious what you think.

Gowsky replied:
It looks good. If TalkBack is disabled, then content descriptions are redundant.

Leedwon replied:
The only drawback is that if you enable touch exploration while being on the screen with chart, this won't update, because `accessibilityManager?.isTouchExplorationEnabled == true` is just an one-off call, we're not observing changes. But I think real TalkBack users probably have it running all the time

Leedwon commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/data/ColumnCartesianLayerModel.kt:
```diff
@@ -180,11 +180,38 @@ public class ColumnCartesianLayerModel : CartesianLayerModel {
+     * descriptions ([contentDescriptions]). All lists should have the same size.
+     */
+    @JvmName("seriesWithContentDescriptions")
+    public fun series(
```
I added new series methods for each chart that take `contentDescriptions: Collection<String?>`. It does add a bit of code bloat, and an alternative would be to just have a nullable `contentDescriptions: Collection<String?>? = null`. But I felt like that might get confusing - both the collection and the strings inside it can be null, so it’s not super clear what a null value actually means. From an end-user/API perspective, I think the current approach is cleaner, even if it adds some overhead internally. But I'm happy to switch to the other approach if you think it's better

Gowsky replied:
Made irrelevant by switching to a provider, as discussed.

Leedwon commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/data/ColumnCartesianLayerModel.kt:
```diff
@@ -180,11 +180,38 @@ public class ColumnCartesianLayerModel : CartesianLayerModel {
+     * Adds a series with the provided _x_ values ([x]), _y_ values ([y]), and optional content
+     * descriptions ([contentDescriptions]). All lists should have the same size.
+     */
+    @JvmName("seriesWithContentDescriptions")
```
`@JvmName` is needed here because the compiler can’t differentiate between `series(Collection<Number>, Collection<Number>)` and `series(Collection<Number>, Collection<Number>, Collection<String?>)`.

Gowsky replied:
Made irrelevant by switching to a provider, as discussed.

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -90,15 +90,17 @@ private fun ColumnCartesianMarkerHighlighter(
-      )
-    }
+    target.columns
+      .sortedBy { it.canvasY }
```
Without `sortedBy`, the tallest column gets drawn on top of the others, which blocks them from being announced by accessibility services. Sorting the data ensures all columns are accessible and get announced properly. You can see this issue in the Daily digital-media use chart.

Ideally, we’d only draw the currently highlighted element - that would solve this issue here and in combo charts as well. But, as mentioned in the Out of Scope section of the PR description, there's no reliable callback for when an element gains accessibility focus.

The [article mentioned](https://eevis.codes/blog/2023-07-24/more-accessible-graphs-with-jetpack-compose-part-1-adding-content/) works around this by adding swipe gestures to the whole chart, letting users move the highlighter manually. But that approach wouldn’t fit a real life library like Vico - it would change core behaviour by introducing swipe gestures for all users.

Our goal is to integrate smoothly with accessibility services, so that users can navigate chart elements using system-level swipe gestures - not the library itself mimicking accessibility interactions.

Leedwon commented:
> Hello, @Leedwon. Thanks for your contribution. Below are some general comments. We'll address your comments individually shortly.
> 
> 1. Feature parity across modules is required, so while this can be merged into a feature branch, support for views and Compose Multiplatform will be needed for it to be released.
> 2. A more flexible and cohesive way of letting consumers specify content descriptions would be a a provider-style interface (similar to `DefaultCartesianMarker.ValueFormatter`).
> 3. Rather than treating each target separately, we should group them by _x_ value, so the user selects a given _x_ value and hears a description of all of the associated entries. This is how `CartesianMarker` works, as well as how accessibility is implemented in the article you've shared.
> 4. The `Highlighter` bounds should be corrected so that they don't extend past the `CartesianLayer` area vertically. Considering point 2, the _y_ coordinates of the top and bottom edges of the `Highlighter` should match those of the top and bottom edges of the `CartesianLayer` area.

Hi @Gowsky, thanks for your reply and the suggested ideas. If feature parity is required, I think it would be better to retarget this to the feature branch, to keep this PR focused on the compose part (could you please create one?). Once the compose part is merged, we’ll also have a clearer idea of what basic accessibility for Vico should look like, so we can port it to Views and Compose Multiplatform. I’ll take a look at the other points soon and try to apply your suggestions.

patrickmichalik commented:
Hello, and thanks for the PR! I’ve created a feature branch and made it the target.

Leedwon commented:
> Hello, @Leedwon. Thanks for your contribution. Below are some general comments. We'll address your comments individually shortly.
> 
> 1. Feature parity across modules is required, so while this can be merged into a feature branch, support for views and Compose Multiplatform will be needed for it to be released.
> 2. A more flexible and cohesive way of letting consumers specify content descriptions would be a a provider-style interface (similar to `DefaultCartesianMarker.ValueFormatter`).
> 3. Rather than treating each target separately, we should group them by _x_ value, so the user selects a given _x_ value and hears a description of all of the associated entries. This is how `CartesianMarker` works, as well as how accessibility is implemented in the article you've shared.
> 4. The `Highlighter` bounds should be corrected so that they don't extend past the `CartesianLayer` area vertically. Considering point 2, the _y_ coordinates of the top and bottom edges of the `Highlighter` should match those of the top and bottom edges of the `CartesianLayer` area.

Hi @Gowsky, before I start improving the code in this PR I wanted to clarify couple of things:

2. Could you help me understand how you envision the API for this provider? I think the tricky part is that accessibility usually requires more context than formatting alone, and I’m not sure how we could provide that.
An example of this is the `AITestScores` chart - we have three datasets tied to a single x value: "Image recognition," "Nuanced-language interpretation," and "Programming." Ideally, the screen reader would say something like:
"Year 2019, Image recognition: 9.52, Nuanced-language interpretation: -100, ..."
If we had an API like `public fun getContentDescription(target: CartesianMarker.Target): CharSequence?`, I’m wondering whether it would be possible to know if a given `Target` corresponds to "Image recognition," "Nuanced-language interpretation," or "Programming." That context would need to be provided somehow.
Passing `contentDescription` during entry creation has the benefit of having that context available upfront, but I also get that you'd prefer to keep the Entries API lean. I can rework this to use a provider-style interface, but I’d really appreciate a starting point or example to better understand how we can pass along that metadata. Also, would this interface belong in the `com.patrykandpatrick.vico.multiplatform.cartesian.marker` package, or some other place would be better for it?
3. I think there are multiple valid ways to announce chart data. The article presents one approach, but I wouldn’t take it as the definitive method. In that example, they have the luxury of merging the content description and formatting the x-axis label together with the values — something that’s not so straightforward in general-purpose library. For instance, iOS charts handle this differently — their screen reader goes through all the values in one dataset first, then proceeds to the next dataset. Here’s a demo of chart accessibility on iOS (with sound on): 

https://github.com/user-attachments/assets/b849ec47-63d6-431b-9554-c67f2b4354b5

3. That being said, I agree that grouping content descriptions by x values is a good option, but I just wanted to share some alternative approaches as well. The only issue I see with grouping by x value is that announcing the x value only once per group might be challenging. To do that we’d likely need to support specifying content descriptions for both x and y values separately. For example, in a candlestick chart, instead of announcing just “twelve hour,” we might want to use something like “midnight” - so content description is not simple a formatted x axis value. So the solution I’d propose is:
- Group all content descriptions by x values
- Merge them so they’re read as one semantic entry (and also highlighted only once)
- Avoid introducing a separate interface for providing the x value’s content description - instead, it would be the consumer’s responsibility to provide the full content description via the provider we discussed in point 2.

For example, in the `AITestScores` case, we should provide content descriptions like:
- "Year 2019, Image recognition: 9.52"
- "Year 2019, Nuanced-language interpretation: -100"

It’s not as smart (since “Year 2019” is repeated), but it’s much simpler from a usage standpoint, and it avoids having two separate providers for content descriptions. WDYT?

4. Good point - I’ll improve that. I guess I should use the height of the `CartesianLayer` rather than the whole `Canvas`, right?

Gowsky commented:
The provider's function wouldn't receive a single `CartesianMarker.Target` but rather a list of them, just like `DefaultCartesianMarker.ValueFormatter.format`. This addresses both concerns.

Regarding point 2, you can use the index of the `CartesianMarker.Target` in the list, usually in combination with [extras](https://www.patrykandpatrick.com/vico/guide/stable/android/core/cartesian-charts/cartesian-chart-model-producer#extras) (such as a list of series labels). Extras are accessed via `CartesianMeasuringContext.extraStore`, so the provider's function should also have a `context` parameter (once again, like `DefaultCartesianMarker.ValueFormatter.format`).

We also plan on adding an `index` property to `CartesianLayer.Entry`. This will make the indices easier to access (without the need for `forEach` or similar), and more importantly, it'll make it easier to link entries to series in cases where there exist _x_ values for which not all series have _y_ values. In the latter case, the index of the `CartesianMarker.Target` in the list can't be used directly, since there are fewer `CartesianMarker.Target`s than series. Currently, for such setups, an extra that helps map a `CartesianMarker.Target` index for a given _x_ value to the index of the corresponding series must be used. `CartesianLayer.Entry.index` would remove the need for this. Now might be a good time to add this property, so feel free to do so if you're interested.

Regarding point 3, yes, there should be a single provider, but it's not just that _x_ and _y_ would be handled together. Instead, since the function would receive a `CartesianMarker.Target` list with all currently highlighted `CartesianMarker.Target`s, the entire text spoken by TalkBack for a single focus event would be returned from a single invocation, as one string. As a result, there's no issue with the repetition of _x_ values. The provider can return any string at all, with or without the _x_ value repeated before every _y_ value. For the example you've given, the function could return e.g. "Year 2019. Image recognition: 9.52. Nuanced-language interpretation: -100."

> Also, would this interface belong in the `com.patrykandpatrick.vico.multiplatform.cartesian.marker` package, or some other place would be better for it?

This can be ignored for now. It would be optimal to move things around a bit more here, converting `CartesianMarker.Target` to something more general (since it's now no longer specific to `CartesianMarker`). However, this may be challenging to achieve in a backward-compatible manner. Let's focus on it when everything else is done.

> I guess I should use the height of the `CartesianLayer` rather than the whole `Canvas`, right?

Yes. You can use `CartesianDrawingContext.layerBounds`.

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -225,23 +226,26 @@ internal fun CartesianChartHostImpl(
       }
 
-      val drawingContext =
+      val context =
```
`Smart cast to 'CartesianDrawingContext' is impossible, because 'drawingContext' is a property that has open or custom gette` hence `val context`

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -165,6 +165,7 @@ internal fun CartesianChartHostImpl(
         remember(chart.layerPadding, model.extraStore) { chart.layerPadding(model.extraStore) },
       pointerPosition = pointerPosition.value,
     )
+  var drawingContext by remember { mutableStateOf<CartesianDrawingContext?>(null) }
```
I'm not sure if we have other way to pass `drawingContext` to `AccessibilityHighlighter`.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/b411a304d5bcca3d646abc3df2a2143a3678f773 - I'm not sure if we need `rememberCartesianDrawingContext` given that `android.graphics.Canvas` is probably not considered stable by compose, but I introduced it to follow the existing convention. Stability could also be checked and if needed `compose_compiler_config` could be added but this is out of scope of this PR

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -40,115 +38,40 @@ import com.patrykandpatrick.vico.core.cartesian.marker.LineCartesianLayerMarkerT
-  Box {
-    target.points.forEach { point ->
+  Box(modifier) {
+    for ((x, targetsGroup) in groupedTargets) {
```
`.forEach` required API level 24 (current min is 21)

Leedwon replied:
Ah you're right, not sure why I assumed that bot are not available, replaced with `forEach`

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChart.kt:
```diff
@@ -58,6 +59,8 @@ public fun rememberCartesianChart(
   decorations: List<Decoration> = emptyList(),
   persistentMarkers: (CartesianChart.PersistentMarkerScope.(ExtraStore) -> Unit)? = null,
   getXStep: ((CartesianChartModel) -> Double) = { it.getXDeltaGcd() },
+  contentDescriptionProvider: DefaultCartesianMarker.ContentDescriptionProvider =
```
I wasn't sure if it should be added to `CartesianChart` or to `CartesianChartHost`. I picked the former, because it is not only compose-specific, but a core class, so the provider can be used in views and multiplatform as well

Leedwon commented on sample/compose/src/main/kotlin/com/patrykandpatrick/vico/sample/compose/GoldPrices.kt:
```diff
@@ -210,21 +210,13 @@ private val high =
     modelProducer.runTransaction {
       // Learn more: https://patrykandpatrick.com/y3c4gz.
-      candlestickSeries(x, opening, closing, low, high, contentDescriptions)
+      candlestickSeries(x, opening, closing, low, high)
```
I've added several fixup! commits to revert changes related to adding contentDescription to chart entries, which are no longer relevant.

I'm keeping these commits for now to maintain visibility during review, but once the PR is approved, I'll rebase and squash the fixups to produce a clean final history.

Leedwon commented on sample/compose/src/main/kotlin/com/patrykandpatrick/vico/sample/compose/AITestScores.kt:
```diff
@@ -47,13 +47,26 @@ import com.patrykandpatrick.vico.core.cartesian.data.CartesianChartModelProducer
 
 private val LegendLabelKey = ExtraStore.Key<Set<String>>()
+private val ContentDescriptionProvider =
+  DefaultCartesianMarker.ContentDescriptionProvider { _, targets ->
```
I wanted to use `val legendLabels = context.extraStore[LegendLabelKey]` instead of `legendLabels = data.keys` to showcase the `extraStore` usage, as you suggested. However, for some reason, the `extraStore` available here doesn't contain that key, which leads to app crashes.

I briefly tried to debug it, and it looks like the key is present in the `extraStore` from `measuringContext` when creating the legend, but it's missing from the `drawingContext` in this part of the code. The same thing happens with the marker `ValueFormatter` - the key is missing during the format call.

Could you help me understand why these extraStores differ, and how we could ensure that the key is available in this context?

Gowsky replied:
>Could you help me understand why these extraStores differ

`ExtraStore` defined in `MeasuringContext` is used for internal drawing data. The `ExtraStore`, which has consumer-provided extras, is accessible via `CartesianMeasuringContext.model.extraStore`.

Leedwon replied:
Thanks for the context, I changed it to use `model.extraStore`

Leedwon commented:
Hi @Gowsky,

I added changes that use the Provider approach that you suggested.

### Couple of things to note:
- `Highlighters` now span the full height of the `CartesianLayer` area. This change was necessary because I discovered that, without it, TalkBack could read chart entries out of order. It prioritizes top-to-bottom reading, so if an entry toward the end had a higher `canvasY` than one at the start, it would be read first. Additionally, entries with y-values close to 0 were sometimes skipped entirely. This update not only improves accessibility behavior but also simplifies the implementation.
- I've removed changes that are no longer relevant in this PR via fixup! commits. This keeps the PR review-friendly (no force pushes with dropped commits), and once it's approved, I’ll rebase and squash the fixups to leave a clean, final commit history.
- I ran into some trouble using `extraStore` for content descriptions. I’d appreciate your help here, more info can be found in this comment: https://github.com/patrykandpatrick/vico/pull/1069#discussion_r2032918147

### Demo:

https://github.com/user-attachments/assets/16f23a0e-99b6-4563-aafa-8619bb3b7a9a

Leedwon commented:
Hi @Gowsky I've addressed latest comments, could you please take a look once you find some time

Gowsky commented:
Hi @Leedwon. Yes, I will when I have the time.

Leedwon commented:
Hi @Gowsky, just following up - any chance you’ve had time to take a look? Sorry to nudge, but this is a blocker for some accessibility work on our side.

Gowsky commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -174,63 +181,89 @@ internal fun CartesianChartHostImpl(
+@Composable
+private fun isTouchExplorationEnabled(): Boolean {
+  val context = LocalContext.current
+  val accessibilityManager =
+    context.getSystemService(Context.ACCESSIBILITY_SERVICE) as? AccessibilityManager
```
This should be wrapped with a `remember` block.
```suggestion
  val accessibilityManager =
    remember { context.getSystemService(Context.ACCESSIBILITY_SERVICE) as? AccessibilityManager }
```

Gowsky commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,94 @@
+
+  Box(
+    modifier =
+      Modifier.offset(x = canvasX.pxToDp() - width / 2, y = canvasTopY.pxToDp())
```
`graphicsLayer` has a better performance than `offset` and works well here.
```suggestion
      Modifier.graphicsLayer {
          translationX = canvasX - width.toPx() / 2
          translationY = canvasTopY
        }
```

Gowsky commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/DefaultCartesianMarker.kt:
```diff
@@ -309,6 +309,22 @@ public open class DefaultCartesianMarker(
+    public fun getContentDescription(
+      context: CartesianDrawingContext,
+      targets: List<CartesianMarker.Target>,
+    ): CharSequence
```
The accessibility services expect `String`, thus returning a `CharSequence` here most likely isn't what we're looking for. Any extra information stored in `CharSequence` subclasses will be lost.

Gowsky commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/DefaultCartesianMarker.kt:
```diff
@@ -385,3 +401,38 @@ internal class DefaultValueFormatter(
+  private fun StringBuilder.append(target: CartesianMarker.Target) {
+    when (target) {
+      is CandlestickCartesianLayerMarkerTarget -> {
+        append("x ${target.x}")
+        append("opening ${target.entry.opening}")
+        append("closing ${target.entry.closing}")
+        append("low ${target.entry.low}")
+        append("high ${target.entry.high}")
```
We mustn't hardcode words, as they won't work for different languages. People might use the `DefaultContentDescriptionProvider` thinking that it might work with different languages. We can turn these words into constructor parameters, so API consumers can provide their own translatable texts.

Gowsky commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/DefaultCartesianMarker.kt:
```diff
@@ -309,6 +309,22 @@ public open class DefaultCartesianMarker(
   }
 
+  /** Provides contentDescription for [CartesianMarker] * */
+  public fun interface ContentDescriptionProvider {
```
This should be moved to a separate file. It isn't used by `DefaultCartesianMarker` and isn't really related to `CartesianMarker` in general.

Gowsky reviewed (changes requested):
There is one more important thing to add. `CartesianLayerModel.Entry` has no `seriesIndex` property which is required for `ContentDescriptionProvider` to label values properly. The content description for AI test scores chart uses "Image recognition" legend label for every top-most point at a given _x_ value.

Gowsky reviewed (changes requested):
>Please note that I’ve defaulted seriesIndex to 0 in public constructors. This isn’t ideal, since we can’t infer a meaningful index outside of the builder scope, but I wanted to avoid introducing a breaking change.

The old public constructors without a `seriesIndex` parameter should be deprecated and use `-1` for the `seriesIndex`. This invalid value can explicitly point out that a deprecated API is being used. New constructors with `seriesIndex` mustn't have any default value.

Gowsky commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/CartesianDrawingContext.kt:
```diff
@@ -66,7 +66,6 @@ internal fun CartesianDrawingContext.getVisibleXRange(): ClosedFloatingPointRang
 @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
 public fun CartesianDrawingContext(
   measuringContext: CartesianMeasuringContext,
-  canvas: Canvas,
```
As per my [previous comment](https://github.com/patrykandpatrick/vico/pull/1069#discussion_r2143354417), `canvas` parameter should remain, but it has to be made nullable. The `view` module should keep using it.

Gowsky commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/data/CandlestickCartesianLayerModel.kt:
```diff
@@ -125,6 +125,8 @@ public class CandlestickCartesianLayerModel : CartesianLayerModel {
    * @property absoluteChange represents the absolute price change ([closing] vs. [opening]).
    * @property relativeChange represents the relative price change (this [Entry]’s [closing] vs. the
    *   previous [Entry]’s [closing]).
+   * @property contentDescription Text used by accessibility services to describe what this entry
```
There's no such property in this class.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/57f2299323565ac4816e45d66d151b9fa27bfa72

Gowsky commented on vico/views/src/main/java/com/patrykandpatrick/vico/views/cartesian/CartesianChartView.kt:
```diff
@@ -336,7 +336,6 @@ constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
       val drawingContext =
         CartesianDrawingContext(
           measuringContext,
-          canvas,
```
This needs to be reverted, as requested in my previous comment. Currently, all charts for the view system cause a crash due to null `Canvas`.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/6211fcd30d83eef8a0f783d52e6e56f28e17dfb9

Gowsky reviewed (changes requested):
Please reformat the code with ktfmt, as per [the contribution guidelines](https://www.patrykandpatrick.com/vico/guide/stable/contributing).

Leedwon commented:
> Please reformat the code with ktfmt, as per [the contribution guidelines](https://www.patrykandpatrick.com/vico/guide/stable/contributing).

Done https://github.com/patrykandpatrick/vico/pull/1069/commits/d0bb28da86fe529ccca0f3f758217a9142acde8a, https://github.com/patrykandpatrick/vico/pull/1069/commits/41b779414398bdd40d5f464b333601f5b1a86025

Gowsky commented on sample/compose/src/main/kotlin/com/patrykandpatrick/vico/sample/compose/AITestScores.kt:
```diff
@@ -47,13 +47,28 @@ import com.patrykandpatrick.vico.core.cartesian.data.CartesianChartModelProducer
+    target.points.forEach { point ->
+      val seriesIndex = point.entry.seriesIndex
+      val label = legendLabels[seriesIndex]
+      append("$label: ${point.entry.y}")
```
Adding a period at the end of this string adds a necessary pause when TalkBack reads the values for each marked year. Without it, the description being read is harder to understand.
```suggestion
      append("$label: ${point.entry.y}.")
```

Gowsky reviewed (changes requested):
There's one last issue I've found when running the sample app. Content descriptions can be improved, so they are easier to understand by adding missing pauses in the speech. I've added a comment with a suggestion for `AITestScores`, but please update other charts so they match the improved behavior.

Leedwon commented:
Hi @Gowsky, I’ve rebased and squashed all fixup commits to keep the history clean. From my side, the PR is ready - unless you see anything that needs changing, I think we can go ahead and merge it.

Leedwon commented:
Hi @patrickmichalik 👋, is there a chance you could take a look sometime soon? Just asking to get an idea of when it might be merged 🙏

Leedwon commented:
Hi @patrickmichalik 👋 Just wanted to check in again on this - totally understand if things are busy, but I’d appreciate any thoughts or an idea of when this might get a look. Thanks so much! 🙏

patrickmichalik commented on sample/compose/src/main/kotlin/com/patrykandpatrick/vico/sample/compose/DailyDigitalMediaUse.kt:
```diff
@@ -57,6 +60,16 @@ private val YDecimalFormat = DecimalFormat("#.## h")
 private val StartAxisValueFormatter = CartesianValueFormatter.decimal(YDecimalFormat)
 private val StartAxisItemPlacer = VerticalAxis.ItemPlacer.step({ 0.5 })
 private val MarkerValueFormatter = DefaultCartesianMarker.ValueFormatter.default(YDecimalFormat)
+private val ContentDescriptionProvider = ContentDescriptionProvider { context, targets ->
+  val legendLabels = context.model.extraStore[LegendLabelKey]
+  val target = targets.first() as ColumnCartesianLayerMarkerTarget
+  buildString {
+    append("Year: ${target.x.toInt()}.")
+    target.columns.zip(legendLabels).forEach { (column, label) ->
+      append("$label: ${column.entry.y}.")
+    }
+  }
+}
```
In sample-chart files, we separate root-level variable definitions with blank lines if any of the definitions is multiline.

patrickmichalik commented on sample/compose/src/main/kotlin/com/patrykandpatrick/vico/sample/compose/ElectricCarSales.kt:
```diff
@@ -40,15 +40,25 @@ import com.patrykandpatrick.vico.core.cartesian.data.CartesianLayerRangeProvider
 import java.text.DecimalFormat
 import kotlinx.coroutines.runBlocking
 
 private val RangeProvider = CartesianLayerRangeProvider.fixed(maxY = 100.0)
 private val YDecimalFormat = DecimalFormat("#.##'%'")
 private val StartAxisValueFormatter = CartesianValueFormatter.decimal(YDecimalFormat)
 private val MarkerValueFormatter = DefaultCartesianMarker.ValueFormatter.default(YDecimalFormat)
+private val ContentDescriptionProvider = ContentDescriptionProvider { _, targets ->
+  val target = targets.first() as LineCartesianLayerMarkerTarget
+  buildString {
+    append("Year: ${target.x.toInt()}.")
+    target.points.forEach { point -> append("${point.entry.y}%.") }
+  }
+}
```
In sample-chart files, we separate root-level variable definitions with blank lines if any of the definitions is multiline.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/8695e858dc80534c9221bbecd09102d1dffca806

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianDrawingContext.kt:
```diff
@@ -0,0 +1,43 @@
+import com.patrykandpatrick.vico.core.cartesian.CartesianMeasuringContext
+import com.patrykandpatrick.vico.core.cartesian.layer.CartesianLayerDimensions
+
+@Composable
+internal fun rememberCartesianDrawingContext(
+  measuringContext: CartesianMeasuringContext,
+  layerDimensions: CartesianLayerDimensions,
+  layerBounds: RectF,
+  scroll: Float,
+  zoom: Float,
+): CartesianDrawingContext {
+  return remember(measuringContext, layerDimensions, layerBounds, scroll, zoom) {
```
An expression body should be used.

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianDrawingContext.kt:
```diff
@@ -0,0 +1,43 @@
+  zoom: Float,
+): CartesianDrawingContext {
+  return remember(measuringContext, layerDimensions, layerBounds, scroll, zoom) {
+    CartesianDrawingContext(
+      measuringContext = measuringContext,
+      layerDimensions = layerDimensions,
+      layerBounds = layerBounds,
+      scroll = scroll,
+      zoom = zoom,
+    )
```
The named-argument syntax is redundant here.

patrickmichalik replied:
Thanks! It looks like you’ve accidentally swapped `scroll` and `zoom`; `scroll` must come first.

Gowsky replied:
The Kotlin style guide specifies in what cases to use named arguments, and this isn’t one of them. A mistake happened, and @patrickmichalik caught it **without** named arguments, so the point doesn’t work.

You could make a similar argument for any overdefensive pattern. If we start treating named arguments as a safety net for inattention, we’re going to end up with noisy code that’s harder to maintain.

Leedwon replied:
Kotlin coding conventions mention that:

> Use the named argument syntax when a method takes multiple parameters of the same primitive type, or for parameters of Boolean type, unless the meaning of all parameters is absolutely clear from context.
[ref](https://kotlinlang.org/docs/coding-conventions.html#named-arguments)

Here a method takes multiple parameters of the same primitive type `scroll` and `zoom`, so I think a case could be made for both approaches.

patrickmichalik replied:
As in the case [discussed below](https://github.com/patrykandpatrick/vico/pull/1069#discussion_r2235217175), since we’re dealing with references to descriptively named variables, the arguments’ meaning is “absolutely clear from context”—making named arguments unnecessary. Here, too, the variables have the same names as the parameters of the construct being invoked, so even if one deemed the arguments’ meaning ambiguous, the use of named arguments would bring no benefit in that respect.

Leedwon replied:
👍 Done https://github.com/patrykandpatrick/vico/pull/1069/commits/b87486fd557bf8f66634d70b4725a46f56667a35

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,97 @@
+  )
+}
+
+@Composable private fun Float.pxToDp() = with(LocalDensity.current) { this@pxToDp.toDp() }
```
A simpler alternative: `LocalDensity.current.run { toDp() }`. (As covered [here](https://kotlinlang.org/docs/scope-functions.html#functions), `run` is preferable to `with` when the lambda’s result is used.)

Gowsky replied:
Here, @patrickmichalik also meant to drop `this@pxToDp`, as it's redundant.

Leedwon replied:
Good point, missed that, done https://github.com/patrykandpatrick/vico/pull/1069/commits/74648e2ae327f5e5f133939c27488dc1dadbf390

patrickmichalik commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/CartesianChart.kt:
```diff
@@ -81,6 +78,10 @@ private constructor(
   private var previousMarkerTargetHashCode: Int?,
   private val persistentMarkerMap: MutableMap<Double, CartesianMarker>,
   private var previousPersistentMarkerHashCode: Int?,
+  /** @suppress */
+  @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+  public val contentDescriptionProvider: ContentDescriptionProvider =
+    ContentDescriptionProvider.default(),
```
I’d place `contentDescriptionProvider` after `getXStep`, so the parameters specific to this constructor remain at the bottom, and those shared with the other one aren’t broken up. Since this constructor is private, backward compatibility isn’t a concern.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/1a998caf9b46d1c8ffcc4a99783cc832c519abab

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,97 @@
+import com.patrykandpatrick.vico.core.cartesian.marker.CartesianMarker
+import com.patrykandpatrick.vico.core.cartesian.marker.ContentDescriptionProvider
+
+/**
+ * Displays accessibility-focused highlight indicators for a set of chart marker targets.
+ *
+ * This function renders invisible highlight elements for each target in the chart to support
+ * accessibility tools like screen readers (e.g., TalkBack). Each marker is rendered with a semantic
+ * `contentDescription`, allowing users to navigate data points via accessibility focus.
+ *
+ * @param targets the list of marker targets representing data points to be highlighted for
+ *   accessibility
+ * @param context holds environment data
+ * @param modifier the modifier to be applied to the [AccessibilityHighlighter].
+ * @param contentDescriptionProvider provides the content description for the highlight elements.
+ */
+@Composable
+internal fun AccessibilityHighlighter(
```
We avoid adding documentation for internal API elements.

patrickmichalik reviewed (changes requested):
Apologies for the wait; I’ve been rather busy. I enclose the first batch of suggestions. I’ll complete my review as soon as I can.

patrickmichalik commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/data/ColumnCartesianLayerModel.kt:
```diff
@@ -169,7 +183,8 @@ public class ColumnCartesianLayerModel : CartesianLayerModel {
      */
     public fun series(x: Collection<Number>, y: Collection<Number>) {
-      series.add(x.zip(y, ColumnCartesianLayerModel::Entry))
+      val nextSeriesIndex = series.size
+      series.add(x.zip(y) { x, y -> Entry(x = x, y = y, seriesIndex = nextSeriesIndex) })
```
I’d rename `nextSeriesIndex` to `seriesIndex`; it concerns the series currently being constructed. Also, for the constructor call, the named-argument syntax is redundant.

Leedwon replied:
I’ll rename it. As for the named arguments, I think they add value when two parameters of the same type are close together, since it helps avoid mistakes like accidentally calling `Entry(y, x, ...)`, which would still compile but be incorrect. That said, it’s a bit of a stylistic preference, so if you prefer the positional style, I’m happy to change it, it’ll work the same either way.

patrickmichalik commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/data/LineCartesianLayerModel.kt:
```diff
@@ -141,7 +155,8 @@ public class LineCartesianLayerModel : CartesianLayerModel {
      */
     public fun series(x: Collection<Number>, y: Collection<Number>) {
-      series.add(x.zip(y, LineCartesianLayerModel::Entry))
+      val nextSeriesIndex = series.size
+      series.add(x.zip(y) { x, y -> Entry(x = x, y = y, seriesIndex = nextSeriesIndex) })
```
I’d rename `nextSeriesIndex` to `seriesIndex`; it concerns the series currently being constructed. Also, for the constructor call, the named-argument syntax is redundant.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/a32077a5d175cdb4f8bbf826d6c1725e7e435efb, https://github.com/patrykandpatrick/vico/pull/1069/commits/d8b9589c852515d9ea2f09f5ef31d68af2d8bc48

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/CartesianChartHost.kt:
```diff
@@ -174,63 +178,86 @@ internal fun CartesianChartHostImpl(
-                    scroll ->
-                    coroutineScope.launch { scrollState.scroll(scroll) }
+  val context =
+    rememberCartesianDrawingContext(
+      measuringContext = measuringContext,
+      layerDimensions = layerDimensions,
+      layerBounds = chart.layerBounds,
+      scroll = scrollState.value,
+      zoom = zoomState.value,
+    )
```
The named-argument syntax is redundant here.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/828210fc4e9b000c6c97f4f4a13539e74a0ce009

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,84 @@
+import com.patrykandpatrick.vico.core.cartesian.marker.CartesianMarker
+import com.patrykandpatrick.vico.core.cartesian.marker.ContentDescriptionProvider
+
+@Composable
+internal fun AccessibilityHighlighter(
+  targets: List<CartesianMarker.Target>,
+  context: CartesianDrawingContext,
+  modifier: Modifier = Modifier,
+  contentDescriptionProvider: ContentDescriptionProvider = ContentDescriptionProvider.default(),
+) {
```
Since `modifier` is unused at `AccessibilityHighlighter`’s only call site, I’d remove it. Also, `contentDescriptionProvider` needn’t have a default value. Once this has been addressed, the named-argument syntax can be dropped at the call site.

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,84 @@
+  modifier: Modifier = Modifier,
+  contentDescriptionProvider: ContentDescriptionProvider = ContentDescriptionProvider.default(),
+) {
+  val groupedTargets = targets.groupBy { it.canvasX }
```
Instead of a lambda, `CartesianMarker.Target::canvasX` can be used.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/75b21d7ceee80df16aa352ec2b2ab7fd5d9dc02f

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,84 @@
+          translationX = canvasX - width.toPx() / 2
+          translationY = canvasTopY
+        }
+        .border(width = borderWidth, color = Color.Transparent)
```
The named-argument syntax is redundant here.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/492686663204a4582f1dfd0ff33cd542bd85768b

patrickmichalik commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,84 @@
+          translationY = canvasTopY
+        }
+        .border(width = borderWidth, color = Color.Transparent)
+        .size(width = width, height = height.pxToDp())
```
The named-argument syntax is redundant here.

Leedwon replied:
Similar to `Entry(x,y)` I think it helps to avoid `.size(height, width)`, but as for the first case I'm happy to remove it.

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/fcb85bff7b707230c3f4ce6c2f9f2fcffadf51e7

patrickmichalik reviewed (changes requested):
Cheers! Here’s the next batch.

patrickmichalik reviewed (changes requested):
Cheers! This should be the penultimate set of comments.

Leedwon commented on vico/compose/src/main/java/com/patrykandpatrick/vico/compose/cartesian/accessibility/AccessibilityHighlighter.kt:
```diff
@@ -0,0 +1,83 @@
+            context = context,
+            targets = targetsGroup,
+          ),
+      )
```
Done, https://github.com/patrykandpatrick/vico/pull/1069/commits/44424e16e110efb61173dac4f4d50a2fc650ae5d

patrickmichalik commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/ContentDescriptionProvider.kt:
```diff
@@ -0,0 +1,69 @@
+    return buildString { targets.forEach { target -> append(target = target) } }
+  }
+
+  private fun StringBuilder.append(target: CartesianMarker.Target) {
```
No worries. Thanks!

patrickmichalik reviewed (changes requested):
Thanks! Some final remarks.

Leedwon commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/ContentDescriptionProvider.kt:
```diff
@@ -0,0 +1,67 @@
+      else -> throw IllegalArgumentException("Unexpected `CartesianMarker.Target` implementation.")
+    }
+  }
+}
```
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/4480c19b76941658655f9fdc36fc2cee283dd642, https://github.com/patrykandpatrick/vico/pull/1069/commits/0ecb66a07ff8c054697fe86bb36d6eea6c14c018

Leedwon replied:
Done https://github.com/patrykandpatrick/vico/pull/1069/commits/cb21941e636cc6d94f7ee3dcefc8df43b4dda08c

Leedwon replied:
Right, appologies for sloppiness latest changes, should be fixed now https://github.com/patrykandpatrick/vico/pull/1069/commits/a646800a4e3aebc2465c124c08d3a72cefb42d27

Leedwon commented:
@patrickmichalik I’ve addressed the latest comments. If there’s nothing else to change, please wait with as I’d like to do an interactive rebase to squash the fixups first

Gowsky commented:
> @patrickmichalik I’ve addressed the latest comments. If there’s nothing else to change, please wait with as I’d like to do a rebase to squash the fixups first. Unless you’re working with squash merging - in which case, feel free to go ahead 🚀.

We will handle the rebase and the squash. Thanks!

Leedwon commented:
Understood, thanks for the reply.
One last question: Would you consider releasing the a11y support for just the Compose part of the library first, or is full Views support a hard requirement before shipping? Releasing the Compose implementation early could already bring value to a11y users.

Gowsky commented:
We’ve always maintained feature parity and want to keep doing so. Furthermore, there were many changes to the `core` module that currently are not in use by the `views` module, which is undesired.
